### ISSUE TITLE
[Feature request]: Sandbox Exemption for V4L2 Hardware Encoders/Decoders

### ISSUE THREAD

**Issue by AbhinavPraveen** [flatpak/flatpak#5524]
### Checklist

- [X] I agree to follow the [Code of Conduct](https://github.com/flatpak/flatpak/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/flatpak/flatpak/issues) for a feature request that matches the one I want to file, without success.

### Suggestion

This is essentially #1715 but for hardware encoders/decoders like that provided by [rkvdec](https://github.com/torvalds/linux/tree/master/drivers/staging/media/rkvdec). 

As, AFAIK, pipewire does not have the ability (or plans to implement the ability) to 'proxy' access to `/dev/video*` devices that are hardware encoders or decoders, sometimes applications may need to access `/dev/video*` directly. A `--devices` group e.g. `--devices=video` or `--devices=vpu` ([VPU](https://en.wikipedia.org/wiki/VPU)) would be nice for these devices as opposed to requiring `--devices=all` in this case. These VPU devices are also usually not hotplugable and so are not dynamic.

These devices (as opposed to all `/dev/video*` devices) could be identified by checking whether the `/dev/video*` device has the [V4L2 M2M Capability](https://www.kernel.org/doc/html/v6.1/userspace-api/media/v4l/dev-mem2mem.html) or perhaps by checking whether the device belongs to a driver in a whitelist e.g. `rkvdec`, `hantro-vpu`, etc.

Related to [GNOME/epiphany#2183](https://gitlab.gnome.org/GNOME/epiphany/-/issues/2183).

**Comment by mcatanzaro** [comment]
We've got basically two options here:

 (a) Add a new `--device=video` permission and mount the video goodies in the sandbox. (Consensus seems to be that video decoders are significantly more likely to be exploited than graphics drivers, and we have a `--device=dri` permission required to access graphics drivers, so it probably shouldn't be available by default.)

This is going to require modifying flatpak-run.c and is probably not too hard. Most likely solution. Then applications would need to use `--device=video` to make it work. Additionally, WebKit's FlatpakLauncher.cpp would pass `--device=video` to flatpak-spawn.

 (b) Design some new mechanism for filtering access to hardware decoders and require that apps use that instead. That is, put a proxy process in the middle that would validate that input passed to the kernel looks sane.

Sebastian's thoughts on this:

> You could probably put a validation layer in between, just needs someone to write that
> Also you need to worry about performance then, and passing stuff efficiently between processes especially. And you need to not just handle v4l2 but also other interfaces 
> Sounds nice, if someone has too much time

**Comment by mcatanzaro** [comment]
Hm, in #1715, the opinion from both @TingPing and @alexlarsson is that video access should go through Pipewire and therefore we should not add `--device=video`. But @philn [says it we cannot use Pipewire for video decoding](https://gitlab.gnome.org/GNOME/epiphany/-/issues/2183#note_1850260). So perhaps the responses in #1715 are not correct?

Additional CC: @alatiera

**Comment by philn** [comment]
If you want to use Pipewire for video decoding you'll need some `pipewirevideodec`  GStreamer element that can be auto-plugged by `decodebin` and similar. Good luck with that!

**Comment by sdroege** [comment]
There's also the aspect here that for non-v4l2 codecs (e.g. NVIDIA and various embedded devices) you need to find a different solution. AFAIU on Android it works by having a service process that the application talks to with a generic API, and that then accesses the hardware in whatever way necessary. We could do the same, but it's a lot of work and I don't see anybody doing that (and supporting all kinds of strange non-v4l2 codecs) unless there's funding available for that.

Until you have an actual plan for that and it's going to be completed in the near future, I'd suggest to simply pass through the specific devices.

**Comment by doraskayo** [comment]
Can camera device nodes under `/dev/video*` also have the aforementioned M2M capability? If so, exposing those would  "accidentally" expose camera device nodes as well, which is a serious privacy concern.

**Comment by AbhinavPraveen** [comment]
> Can camera device nodes under `/dev/video*` also have the aforementioned M2M capability? If so, exposing those would "accidentally" expose camera device nodes as well, which is a serious privacy concern.

According to [these docs](https://www.kernel.org/doc/html/v6.1/userspace-api/media/v4l/open.html#related-devices), camera nodes and M2M nodes should be distinct.

> Devices can support several functions. For example video capturing, VBI capturing and radio support.
>
> The V4L2 API creates different V4L2 device nodes for each of these functions.
>
> The V4L2 API was designed with the idea that one device node could support all functions. However, in practice this never worked: this ‘feature’ was never used by applications and many drivers did not support it and if they did it was certainly never tested. In addition, switching a device node between different functions only works when using the streaming I/O API, not with the read()/write() API.
>
> Today each V4L2 device node supports just one function.

Although someone more familiar with the kernel API is welcome to comment.

We can also check if the device belongs to a driver in a whitelist (using e.g. [this](https://www.kernel.org/doc/html/v6.1/userspace-api/media/v4l/vidioc-querycap.html)) which would also prevent an accidental exposure.

**Comment by AbhinavPraveen** [comment]
Of course, we could also exclude `/dev/video*` devices that have the [Video Capture](https://www.kernel.org/doc/html/v6.1/userspace-api/media/v4l/dev-capture.html) capability or any/all other capabilities.

**Comment by alatiera** [comment]
> The downside is that all applications that support video playback are going to need to add this new permission.

Applications that want to access video decoders already need to add a permission, so if we grouped a couple of devices under a new one it should be easy to transition the permissions over, however we also don't get much from it other than slightly better metadata. 

There are `v4l2`, `vappi`, and nvidia specific devices you will need depending on the hardware you use so exposing just v4l2 devices would be incomplete.

**Comment by mcatanzaro** [comment]
> Applications that want to access video decoders already need to add a permission

What permission? This is news to me. I don't see any video-related permission's in Epiphany's manifest, for example.

> There are v4l2, vappi, and nvidia specific devices you will need depending on the hardware you use so exposing just v4l2 devices would be incomplete.

I guess they should all be exposed?

**Comment by alatiera** [comment]
> What permission? This is news to me. I don't see any video-related permission's in Epiphany's manifest, for example.

Yes, cause applications either don't use hw encoders or already have the required device. For example in my laptop `vah264enc` will try to open the device under `/dev/dri/renderD128`.

**Comment by mcatanzaro** [comment]
> Applications that want to access video decoders already need to add a permission, so if we grouped a couple of devices under a new one it should be easy to transition the permissions over, however we also don't get much from it other than slightly better metadata.

Well better metadata is important? If you add `--filesystem` permissions then you're going to be displayed as unsafe in GNOME Software. And I refuse to add such permissions to WebKit, for example. Adding a `-device=video` permission avoids this problem.

**Comment by mcatanzaro** [comment]
In https://github.com/flathub/org.gnome.Epiphany/pull/64 we have a user reporting that they got hardware decoding working with Intel graphics even without access to any /dev/video devices. Is v4l2 device access only required for specialty hardware?

**Comment by sdroege** [comment]
Intel would go via VA, and for that you need either a display connection or access to the DRI devices. That's why I said multiple times that V4L2 is only one of the many options (and AFAIU that's only used on embedded Linux for now) :)

**Comment by mcatanzaro** [comment]
Well it's a surprise to me, because it wasn't mentioned yet in this issue report.

My understanding is that VA is the standard expected API for hardware decoding nowadays? So if it works without requiring a new `--dev=video` permission, that's quite important to know.

Maybe enumerating the various possibilities of how hardware decoding is expected to work would help, to avoid missing anything. Because so far all we know from this issue is (a) V4L2 hardware decoders require specific devices in /dev/video to be mounted, (b) VA decoders work fine already with just `--device=dri`, (c) NVIDIA possibly requires something different. Anything else?

**Comment by sdroege** [comment]
> My understanding is that VA is the standard expected API for hardware decoding nowadays

... on desktop Linux and not with NVIDIA GPUs (when using the proprietary driver).

> Anything else?

Various platform-specific things on embedded Linux.

I would suggest covering V4L2, VA and NVIDIA for now. Make sure that actually works. And then make that covered by something like `--device=videocodecs`. And if needed that can be extended later with more things.

**Comment by AbhinavPraveen** [comment]
> Well it's a surprise to me, because it wasn't mentioned yet in this issue report.

Indeed, this issue was originally for hardware encoders/decoders accessed through a V4L2 device like that provided by [rkvdec](https://github.com/torvalds/linux/tree/master/drivers/staging/media/rkvdec). There is also VAAPI which is used by Intel and AMD drivers, VDPAU which is supported by NVidia, Nvenc/Nvdec also supported by Nvidia and potentially more. Of these, indeed VAAPI is the best supported by applications but, I believe, Gstreamer Bad has support for VDPAU and Nvenc/NVdec (and V4L2).

**Comment by sdroege** [comment]
VDPAU is outdated and not supported by GStreamer anymore. NVIDIA has its own proprietary APIs now, which are supported by GStreamer.

**Comment by kepstin** [comment]
> (b) Design some new mechanism for filtering access to hardware decoders and require that apps use that instead. That is, put a proxy process in the middle that would validate that input passed to the kernel looks sane.
> 
> @sdroege's thoughts on this:
> 
> > You could probably put a validation layer in between, just needs someone to write that
> > Also you need to worry about performance then, and passing stuff efficiently between processes especially. And you need to not just handle v4l2 but also other interfaces
> > Sounds nice, if someone has too much time

I find this idea pretty interesting - it would basically end up being a daemon that provides a "hardware codec" portal interface, which would provide the ability to enumerate available hardware codec implementations and capabilities, then request and configure a decoding/encoding session. While pipewire itself doesn't provide access to hardware codec devices, I think it would make sense to use pipewire as the *transport* between the application and the hardware codec daemon. This provides an interesting possibility where the application could link a pipewire stream from camera or stream capture source directly to a hardware decoder or encoder.

Tricky thing is probably mostly in how to design the codec enumeration and configuration apis with enough flexibility to support multiple codec backend implementations without being too difficult to use or implementation-specific. Looking at the [Android MediaCodecList](https://developer.android.com/reference/android/media/MediaCodecList) and [Apple VideoToolbox](https://developer.apple.com/documentation/videotoolbox) APIs might be interesting for reference.

The daemon itself would probably best be implemented either on top of gstreamer or on top of ffmpeg, to take advantage of their work dealing with various interfaces and implementation quirks.

**Comment by rmader** [comment]
Hi, I wanted to shortly outline that there's an increasing number of apps/platforms where the original requests would be beneficial. One simple example is gnome-camera / snapshot on ARM devices - it uses the camera via pipewire/portal and, as a recording app, would want to have access V4L2 hw encoders. The same goes for apps like Dino (xmpp client with video chat support) - it needs hw encoding and decoding in order to have acceptable performance on e.g. linux phones (and not drain the battery way too fast). In the long run I'd expect most apps making use of the camera portal will want to have access to hw-encoders - apart from those only needing it for e.g. QR-code scanning.

Given the static nature of V4L2 de-/encoders (hot-plug support really not needed) I think the original proposal makes sense - and is also what e.g. [Firefox does](https://phabricator.services.mozilla.com/D178194). One thing to add there 
is that there's the stateful (m2m) and the stateless API, so only checking for m2m is not enough.

Regarding proposals around Pipewire codec support: I like the idea as well, but think that it would take years to realize - it's neither trivial, nor do I see a strong case for commercial users to invest much into it atm. I.e. it's similar to Vulkan video in that it's nice - but the advantage over just supporting VA-API and V4L2 is not very obvious. Therefore it's IMO not a solution for the conceivable future.

So if there is no strong resistance against a static filtering solution for `/dev/video*` devices, I'd look into porting the FF solution - WDYT? :)

P.S.: Given the very specific nature I wonder if `--device=v4l2codecs` would be a fitting name - as said before, VA-API (and probably some upcoming VK implementations?) are already covered by `--device=dri` so it probably makes sense to be more specific here.

**Comment by Erick555** [comment]
> So if there is no strong resistance against a static filtering solution for /dev/video* devices, I'd look into porting the FF solution - WDYT? :)

on several occasions flatpak developers [stance](https://github.com/flatpak/flatpak/issues/4405) was that exposing non-subdirectory under /dev is no-go due to lack of plug&play scenario support

**Comment by rmader** [comment]
> due to lack of plug&play scenario support

While that's certainly important for USB, it's AFAIK a practically non-existent scenario for V4L2 en-/decoders - much less even than for VA-API, where plug&play of GPUs is at least somewhat realistic. That's why I think it shouldn't be a blocker here.

**Comment by Erick555** [comment]
This [comment](https://github.com/flatpak/flatpak/issues/4405#issuecomment-917650757) specifically uses `/dev/video*` example. I guess plugging in/out usb webcam isn't something unthinkable. Personally I do /dev filtering with pure bubblewrap (which is what flatpak uses under the hood) and even suspend/resume cycle breaks my devices in such case for reason I not yet fully understand.

Anyway you should consult flatpak developers on this before implementing it somewhere.

**Comment by rmader** [comment]
> This https://github.com/flatpak/flatpak/issues/4405#issuecomment-917650757 specifically uses /dev/video* example. I guess plugging in/out usb webcam isn't something unthinkable.

Right, but while runtime plugging of webcams is something that we would want to support (and is commonly supported by apps), the same is not really the case here, since video en-/decoders are usually permanent (notable exception: `vivid` devices for testing/debugging, only relevant for devs). But yeah, this needs careful evaluation. Will try to catch somebody to talk this through.

**Comment by hfiguiere** [comment]
Before adding a new device we need support for backward compatibility, otherwise, like device=input (added for #5481), it will be useless since the only way to support flatpak < 1.15.6 is to add device=all.

**Comment by rmader** [comment]
> Before adding a new device we need support for backward compatibility, otherwise, like device=input (added for https://github.com/flatpak/flatpak/pull/5481), it will be useless since the only way to support flatpak < 1.15.6 is to add device=all.

While I agree that it would be good to have, hardware en/-decoding is often not critical (since software fallbacks exist). So in absence of that issue being solved, I'd expect app developers to be fine with a solution that only gives users with recent Flatpak the optimal experience. This is evident by the fact that right now many affected apps - even dedicated video players - ship without `device=all`, reducing video decoding performance significantly on ARM devices.

**Comment by rmader** [comment]
@smcv @alexlarsson When you have time, can I ask you to shortly state whether you would have strong objections against this getting implemented?

TL;DL is: v4l2 video en-/decoders are persistent/non-hotplugable devices and it would be very helpful for multimedia apps to be able to use them without opening the sandbox too widely with `device=all`. They share the naming scheme with (hotplugable) v4l2 cameras, however it's possible to iterate over all `/dev/video*` devices and mount en-/decoders to containers at startup.

**Comment by ndufresne** [comment]
Please, reminder that for stateless decoders, there is an extra /dev/mediaX device that needs to be available. Software like GStreamer will require access to udev daemon too.

**Comment by rmader** [comment]
> Please, reminder that for stateless decoders, there is an extra /dev/mediaX device that needs to be available.

Uh, good point - but that should be possible with the same approach AFAICS. 

> Software like GStreamer will require access to udev daemon too.

AFAIK Gstreamer works fine `--devices=all` permissions - so that shouldn't be a issue? Or do i miss something here?
